awx/main/models/unified_jobs.py

```diff
@@ -1229,6 +1229,23 @@ def pre_start(self, **kwargs):
             self.save(update_fields=['job_explanation'])
             return (False, None)
 
+        # verify that any associated credentials aren't missing required field data
+        missing_credential_inputs = []
+        for credential in self.credentials.all():
+            defined_fields = credential.credential_type.defined_fields
+            for required in credential.credential_type.inputs.get('required', []):
+                if required in defined_fields and not credential.has_input(required):
+                    missing_credential_inputs.append(required)
```

Comment:
not a blocker, just want to throw it out there - `missing_credential_inputs` might need some visibility from the credential API at some point. That may be a problem for another release.

Reply:
Yea, I _really_ feel like this is an edge case.  Part of me thought about not solving this at all, because if the type says it's required, and you never set a value, that's *kind of* on you.